samples/security/TokenLoginModule.java

```diff
@@ -0,0 +1,256 @@
+
+        try {
+
+            AuthenticationToken authToken = getTokenFromCallback();
+            if ((authToken == null) || (authToken.isEmpty())) {
+                // Since a token was expected, issue a challenge to the client to provide one.
+                // This will result in a 401 HTTP response back to the client.
+                LoginResult loginResult = getLoginResultFromCallback();
+                loginResult.challenge();
+
+                // NOTE: Optionally, the login module can send data to the client
+                // along with the challenge. The following commented out line shows
+                // an example of how to do that.
+                //
+                // On the client side, this data will be included in the challenge
+                // request and can be extracted by the client to be used in any way.
+                //
+                // loginResult.challenge("app-id=123456789");
+
+                throw new LoginException("No tokens provided");
+            }
+            debug(String.format("Login: Obtained AuthenticationToken: %s", authToken));
+
+            // Ensure that the token data uses the "Token" authentication scheme.
+            // Note: Compare to "Token" rather than "Application Token" because the scheme
+            // is just the scheme type which is Basic, Negotiate, or Token.
+            String authScheme = authToken.getScheme();
+            if (!authScheme.equals("Token")) {
+                throw new LoginException(
+                    "Expected authentication token scheme 'Token', but received '" + authToken.getScheme() + "'");
+            }
+
+            // The AuthenticationToken contains extra information (such as the scheme), so
+            // extract the token data specifically.
+            String tokenData = authToken.get();
+
+            processToken(tokenData);
+
+            return true;
+
+        }
+        catch (LoginException e) {
+            debug(String.format("Login: Got a LoginException: %s", e.getMessage()));
+            throw e;
+        }
+        catch (Exception e) {
+            debug(String.format("Login: Got an Exception: %s", e.getMessage()));
+            throw new LoginException(e.getMessage());
+        }
+
+    }
+
+    @Override
+    public boolean commit() throws LoginException {
+        // Save the username into shared state for downstream login modules to authorize.
+        debug(String.format("Commit: Saving username to shared state: %s", username));
+
+        sharedState.put(USERNAME_KEY, username);
+
+        if (debug) {
+            if (sharedState.size() > 0) {
+                debug("Commit: Shared state:");
+                sharedState.forEach((k, v) -> debug(String.format("Commit:   %s=%s", k, v)));
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean abort() throws LoginException {
+        // Clear state here, if any was created, particularly anything stored in
```

Comment:
Yes. With the realization that the shared state should be set in `login()` rather than `commit()`, then it is correct to clear that shared state here in `abort()`. I will include that in the next commit.